File: typescript_lexer/src/FileReader.ts
import { promises as fs } from 'fs';
import { join } from 'path';

/**
 * Asynchronous file reader utility
 * Demonstrates async/await functionality
 */
export class FileReader {
  /**
   * Asynchronously reads a file and returns its contents
   */
  static async readFile(filePath: string): Promise<string> {
    try {
      const content = await fs.readFile(filePath, 'utf-8');
      return content;
    } catch (error) {
      if (error instanceof Error) {
        throw new Error(`Failed to read file ${filePath}: ${error.message}`);
      }
      throw error;
    }
  }

  /**
   * Asynchronously reads multiple files
   * Demonstrates parallel async operations
   */
  static async readMultipleFiles(filePaths: string[]): Promise<Map<string, string>> {
    const results = new Map<string, string>();

    // Read all files in parallel using Promise.all
    const filePromises = filePaths.map(async (filePath) => {
      const content = await FileReader.readFile(filePath);
      return { filePath, content };
    });

    const fileResults = await Promise.all(filePromises);

    // Store results in a map
    fileResults.forEach(({ filePath, content }) => {
      results.set(filePath, content);
    });

    return results;
  }

  /**
   * Gets the full path for an example file
   */
  static getExamplePath(filename: string): string {
    return join(__dirname, '..', 'examples', filename);
  }

  /**
   * Lists all example files asynchronously
   */
  static async listExampleFiles(): Promise<string[]> {
    const examplesDir = join(__dirname, '..', 'examples');
    try {
      const files = await fs.readdir(examplesDir);
      return files.filter(file => file.endsWith('.txt'));
    } catch (error) {
      if (error instanceof Error) {
        throw new Error(`Failed to list example files: ${error.message}`);
      }
      throw error;
    }
  }
}

File: erlang_lexer/src/file_reader.erl
-module(file_reader).
-export([read_file/1, read_multiple_files/1, list_example_files/0, get_example_path/1]).

%% Read a single file and return its contents as a string.
-spec read_file(string()) -> {ok, string()} | {error, term()}.
read_file(FilePath) ->
    case file:read_file(FilePath) of
        {ok, Binary} ->
            {ok, binary_to_list(Binary)};
        {error, Reason} ->
            {error, {file_error, FilePath, Reason}}
    end.

%% Read multiple files concurrently using spawned processes.
%% Returns a list of {Filename, Contents} tuples in the original order.
-spec read_multiple_files([string()]) -> [{string(), {ok, string()} | {error, term()}}].
read_multiple_files(FilePaths) ->
    Parent = self(),
    Refs = lists:map(
        fun(Path) ->
            Ref = make_ref(),
            spawn(fun() ->
                Result = read_file(Path),
                Parent ! {Ref, Path, Result}
            end),
            {Ref, Path}
        end,
        FilePaths
    ),
    collect_results(Refs, []).

%% Collect results from spawned file-reading processes in order.
collect_results([], Acc) ->
    lists:reverse(Acc);
collect_results([{Ref, Path} | Rest], Acc) ->
    receive
        {Ref, Path, Result} ->
            collect_results(Rest, [{Path, Result} | Acc])
    after 5000 ->
        collect_results(Rest, [{Path, {error, timeout}} | Acc])
    end.

%% Construct the path to an example file.
-spec get_example_path(string()) -> string().
get_example_path(Filename) ->
    filename:join([get_examples_dir(), Filename]).

%% List all .txt files in the examples directory.
-spec list_example_files() -> {ok, [string()]} | {error, term()}.
list_example_files() ->
    Dir = get_examples_dir(),
    case file:list_dir(Dir) of
        {ok, Files} ->
            TxtFiles = [F || F <- Files, filename:extension(F) =:= ".txt"],
            {ok, lists:sort(TxtFiles)};
        {error, Reason} ->
            {error, {dir_error, Dir, Reason}}
    end.

%% Get the path to the examples directory relative to this module.
get_examples_dir() ->
    ScriptDir = filename:dirname(code:which(?MODULE)),
    filename:join([filename:dirname(ScriptDir), "examples"]).
